ingester-protocol/src/main/java/io/greptime/RouterClient.java

```diff
@@ -57,19 +65,44 @@ public boolean init(RouterOptions opts) {
         List<Endpoint> endpoints = Ensures.ensureNonNull(this.opts.getEndpoints(), "null `endpoints`");
 
         this.router = new DefaultRouter();
-        this.router.onRefresh(endpoints);
+        this.router.onRefresh(endpoints, null);
 
         long refreshPeriod = this.opts.getRefreshPeriodSeconds();
         if (refreshPeriod > 0) {
             this.refresher = REFRESHER_POOL.getObject();
             this.refresher.scheduleWithFixedDelay(
-                    () -> this.router.refresh().whenComplete((r, e) -> {
-                        if (e != null) {
-                            LOG.error("Router cache refresh failed.", e);
-                        } else {
-                            LOG.debug("Router cache refresh {}.", r ? "success" : "failed");
-                        }
-                    }),
+                    () -> {
+                        long thisSequence = this.refreshSequencer.incrementAndGet();
+                        checkHealth().whenComplete((r, t) -> {
+                            if (t != null) {
+                                LOG.warn("Failed to check health", t);
+                                return;
+                            }
+
+                            synchronized (this) {
```

Comment:
Is this `this` refer to the closure?

Reply:
this refer to the RouterClient instance

Reply:
I changed the lock to `this.refreshSequencer`; it has the same effect but is easier to understand.